extensions/amp-insticator/0.1/amp-insticator.js

```diff
@@ -0,0 +1,179 @@
+ * limitations under the License.
+ */
+
+
+
+/* ++++++++++ --------------- IMPORTS --------------- ++++++++++ */
+import { Layout } from '../../../src/layout';
+import { setStyles } from '../../../src/style';
+
+
+
+/* ++++++++++ --------------- EXPORT AMP-INSTICATOR --------------- ++++++++++ */
+export class AmpInsticator extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iFrameElement = null; // amp requirement
+
+    /** @private {!Object} */
+    this.url = {
+      ads: 'https://drhn9v8cwg89y.cloudfront.net',
+      embed: 'https://d3lcz8vpax4lo2.cloudfront.net'
+    }
+  }
+
+
+  // ~~~~~ AMP METHODs ~~~~~ //
+  /**
+   * @param {boolean=} onLayout
+   * @override
+   */
+  preconnectCallback(onLayout) {
+    // define preconnect
+    const { preconnect } = this;
+    // host where we store advertising settings code
+    preconnect.url(this.url.ads);
+    // host we serve the core embed application from
+    preconnect.url(this.url.embed);
+  }
+
+  /** @override */
+  buildCallback() {
+    // create markup structure
+    const template = this.createTemplate(this.createInitialMarkup());
+    // append markup structure to DOM
+    this.appendElement(this.element, template);
+  }
+
+  /** @override */
+  layoutCallback() {
+    // get data attribute from the amp-insticator tag
+    const embedId = this.element.getAttribute('data-embed-id');
+    // store DOM elements
+    const insticatorContainer = this.element.querySelector('#insticator-container');
+    const embedIframe = this.iFrameElement = this.element.querySelector('#insticator-iframe');
+    const embedApp = this.createElement(this.element.ownerDocument, 'div', { id: 'app' });
+    const embedScript = this.createElement(this.element.ownerDocument, 'script', { type: 'text/javascript', src: `${this.url.embed}/embed-code/${embedId}.js` });
+
+    // append iframe
+    this.appendElement(this.element.querySelector('#insticator-embed'), embedIframe);
+
+    // append content to iframe
+    this.appendElement(embedIframe.contentWindow.document.body, embedApp);
+    this.appendElement(embedIframe.contentWindow.document.head, embedScript);
+
+    // append ads
+    this.getRequest(`${this.url.ads}/test/ad_settings/${embedId}.js`, (ads) => this.appendAds(ads));
+
+    // apply custom styles
+    setStyles(insticatorContainer, { 'text-align': 'center' });
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iFrameElement) {
+      removeElement(this.iFrameElement);
+      this.iFrameElement= null;
+    }
+    return true; // Call layoutCallback again.
+  }
+
+
+  // ----- CUSTOM FUNCTIONs ----- //
+  /**
+   * Make get request to get ads data.
+   * @param {!string} file
+   * @param {function(Object)} callback
+   */
+  getRequest(file, callback) {
```

Comment:
what do ads data look like?

Reply:
Below is our current dummy data. The properties in these objects will be different in the final implementation. 
`{
	"ad1": {
		"width": "300",
		"height": "250",
		"type": "a9",
		"data-amzn_assoc_ad_mode": "auto",
		"data-divid": "amzn-assoc-ad-fe746097-f142-4f8d-8dfb-45ec747632e5",
		"data-recomtype": "async",
		"data-adinstanceid": "fe746097-f142-4f8d-8dfb-45ec747632e5"
	},
	"ad2": {
		"width": "300",
		"height": "250",
		"type": "a9",
		"data-amzn_assoc_ad_mode": "auto",
		"data-divid": "amzn-assoc-ad-fe746097-f142-4f8d-8dfb-45ec747632e5",
		"data-recomtype": "async",
		"data-adinstanceid": "fe746097-f142-4f8d-8dfb-45ec747632e5"
	}
}`